borgmatic/hooks/zabbix.py

```diff
@@ -0,0 +1,93 @@
+import logging
+import requests
+import json
+
+logger = logging.getLogger(__name__)
+
+
+def initialize_monitor(
+    ping_url, config, config_filename, monitoring_log_level, dry_run
+):  # pragma: no cover
+    '''
+    No initialization is necessary for this monitor.
+    '''
+    pass
+
+
+def ping_monitor(hook_config, config, config_filename, state, monitoring_log_level, dry_run):
+    '''
+    Ping the configured Zabbix topic. Use the given configuration filename in any log entries.
+    If this is a dry run, then don't actually ping anything.
+    '''
+
+    run_states = hook_config.get('states', ['fail'])
+
+    if state.name.lower() in run_states:
+        dry_run_label = ' (dry run; not actually pinging)' if dry_run else ''
+
+        state_config = hook_config.get(state.name.lower(),{'value': f'invalid',},)
+
+        base_url = hook_config.get('server', 'https://cloud.zabbix.com/zabbix/api_jsonrpc.php')
+        username = hook_config.get('username')
+        password = hook_config.get('password')
+        api_key = hook_config.get('api_key')
+        itemid = hook_config.get('itemid')
+        host = hook_config.get('host')
+        key = hook_config.get('key')
+
+        value = state_config.get('value')
+
+        headers = {'Content-Type': 'application/json-rpc'}
+
+        logger.info(f'{config_filename}: Pinging zabbix {dry_run_label}')
+        logger.debug(f'{config_filename}: Using zabbix ping URL {base_url}')
+
+        # Determine the zabbix method used to store the value: itemid or host/key
+        if (itemid) is not None:
+            logger.info(f'{config_filename}: Updating {itemid} on zabbix')
+            data = {"jsonrpc":"2.0","method":"history.push","params":{"itemid":itemid,"value":value},"id":1}
+        
+        elif (host and key) is not None:
+            logger.info(f'{config_filename}: Updating Host:{host} and Key:{key} on zabbix')
+            data = {"jsonrpc":"2.0","method":"history.push","params":{"host":host,"key":key,"value":value},"id":1}
+
+        elif (host) is not None:
+            logger.warning( f'{config_filename}: Key missing for zabbix authentication' )
+
+        elif (key) is not None:
+            logger.warning( f'{config_filename}: Host missing for zabbix authentication' )
+
+        # Determine the authentication method: API key or username/password
+        auth = None
+        if (api_key) is not None:
+            logger.info(f'{config_filename}: Using API key auth for zabbix')
+            headers['Authorization'] = 'Bearer ' + api_key
+            
+        elif (username and password) is not None:
+            logger.info(f'{config_filename}: Using user/pass auth with user {username} for zabbix')
+            response = requests.post(base_url, headers=headers, data='{"jsonrpc":"2.0","method":"user.login","params":{"username":"'+username+'","password":"'+password+'"},"id":1}')
```

Comment:
Code style convention: Use [Python f-strings](https://docs.python.org/3/tutorial/inputoutput.html#formatted-string-literals) here instead of interpolating variables with string concatenation.

Reply:
done: https://github.com/borgmatic-collective/borgmatic/pull/85/commits/b42793a2dc8a3b8b60c3c02ea78924c17955cc21